packages/aws-cdk/lib/api/hotswap-deployments.ts

```diff
@@ -59,13 +66,58 @@ export async function tryHotswapDeployment(
 
 async function findAllHotswappableChanges(
   stackChanges: cfn_diff.TemplateDiff, evaluateCfnTemplate: EvaluateCloudFormationTemplate,
+  sdk: ISDK, rootStackArtifact: cxapi.CloudFormationStackArtifact, nestedStackNames: {[key: string]: any}, hotswappableResources?: HotswapOperation[],
 ): Promise<HotswapOperation[] | undefined> {
   const resourceDifferences = getStackResourceDifferences(stackChanges);
 
   let foundNonHotswappableChange = false;
   const promises: Array<Array<Promise<ChangeHotswapResult>>> = [];
+
+  if (!hotswappableResources) {
+    hotswappableResources = new Array<HotswapOperation>();
+  }
+
+  // process any resources in nested stacks
+  // if any of them are not hotswappable, fail now
+  const resourceDifferenceEntries = Object.entries(resourceDifferences);
+  for (const [logicalId, change] of resourceDifferenceEntries.filter(
+    ([_, resourceDifference]) => (resourceDifference.newValue?.Type === 'AWS::CloudFormation::Stack' && resourceDifference.oldValue?.Type === 'AWS::CloudFormation::Stack'))
+  ) {
+    const parameters = change.newValue?.Properties?.Parameters;
+    if (parameters) {
+      for (const paramName in parameters) {
+        if (typeof parameters[paramName] !== 'string') {
+          parameters[paramName] = await evaluateCfnTemplate.evaluateCfnExpression(parameters[paramName]);
+
+        }
+      }
+    }
+
+    const nestedStackName = nestedStackNames[logicalId].stackName;
+    const evaluateNestedCfnTemplate = nestedStackName
+      ? evaluateCfnTemplate.createNestedEvaluateCloudFormationTemplate(nestedStackName,
+        sdk, rootStackArtifact, parameters)
+      : evaluateCfnTemplate;
+
+    const nestedDiff = cfn_diff.diffTemplate(change.oldValue?.Properties?.NestedTemplate, change.newValue?.Properties?.NestedTemplate);
+    if (await findAllHotswappableChanges(nestedDiff, evaluateNestedCfnTemplate, sdk,
+      rootStackArtifact, nestedStackNames[logicalId].children, hotswappableResources) === undefined
+    ) {
+      return undefined;
+    }
+  }
+
+  // if a stack has been deleted, fail now
+  for (const [_, change] of resourceDifferenceEntries) {
+    if (change.newValue?.Type !== 'AWS::CloudFormation::Stack' && change.oldValue?.Type === 'AWS::CloudFormation::Stack') {
```

Comment:
the comment needs to be updated, this actually checks something more specific; if a nested stack has been deployed to CloudFormation with logical ID `my-id`, and we see a resource with id `my-id` and a type that is not `AWS::CloudFormation::Stack`, we catch that here and return. If we don't do that here, our logic below will not perform a full deployment, and can actually find a `HotswapOperation`: check the test with the description

"attempting to hotswap a newly created resource with the same logical ID that a nested stack which is currently deployed to CloudFormation has results in a full deployment"

to see the test case for this.

Reply:
maybe this would be better expressed as

`if (change.newValue?.Type !== change.oldValue?.Type)`

Reply:
But that means there will be a change to add the new resource with `my-id` in the diff?

Basically, I'm not sure I understand what makes Nested Stacks different than any other resource changing its type, compared to what's currently deployed.

Reply:
That change actually found bugs in our existing tests, which I've just updated

Reply:
I think you're right, there's nothing special about nested stacks here. But currently changing resource types isn't something we accounted for in some cases. eg look at this test: https://github.com/aws/aws-cdk/blob/master/packages/aws-cdk/test/api/hotswap/hotswap-deployments.test.ts#L125

the type changes, but the test succeeds, even though it should fail.

Reply:
Did you link to the correct line of code? Because that's a test about the Metadata resource (`'changes only to CDK::Metadata result in a noOp'`), which I don't think is the problem here 🙂.

Reply:
yep, that's the right test. The type changes from `AWS::CDK::MetaData` to `AWS::CDK::Metadata`

Reply:
Pretty sure that's a typo in the test, and we should have a separate test for this case 🙂.